p2p/p2p.go

```diff
@@ -139,98 +142,70 @@ func (s *Service) sendStatusMessages(peer peer.ID) {
+func (s *Service) broadcastMessages() {
 	for {
-		// receive message from core service (from BABE session)
+		// receive message from core service
 		msg := <-s.msgRec
 
 		log.Trace(
-			"Received message",
-			"host", s.host.id(),
+			"Broadcasting message from core service",
+			"host", s.ID(),
+			"type", msg.GetType(),
 		)
 
-		// check if message should be broadcasted
-		if !s.shouldBroadcast(msg) {
-			log.Trace(
-				"Message ignored",
-				"host", s.host.id(),
-				"type", msg.GetType(),
-			)
-			return
-		}
-
-		// send message to each connected peer
+		// broadcast message to connected peers
 		s.host.broadcast(msg)
 	}
 }
 
-// shouldBroadcast checks if message is new with a valid type, storing the
-// result for later checks and returning true if its a valid new message
-func (s *Service) shouldBroadcast(msg Message) bool {
-	msgType := msg.GetType()
-
-	switch msgType {
-	case BlockRequestMsgType:
-		if s.blockReqRec[msg.Id()] {
-			return false
-		}
-		s.blockReqRec[msg.Id()] = true
-	case BlockResponseMsgType:
-		if s.blockRespRec[msg.Id()] {
-			return false
-		}
-		s.blockRespRec[msg.Id()] = true
-	case BlockAnnounceMsgType:
-		if s.blockAnnRec[msg.Id()] {
-			return false
-		}
-		s.blockAnnRec[msg.Id()] = true
-	case TransactionMsgType:
-		if s.txMessageRec[msg.Id()] {
-			return false
-		}
-		s.txMessageRec[msg.Id()] = true
-	default:
-		// status message type not valid
-		return false
-	}
-
-	return true
-}
-
 // handleStream parses the message written to the data stream and calls the
 // associated message handler (status or non-status) based on message type
 func (s *Service) handleStream(stream net.Stream) {
 
-	log.Trace(
-		"Received message",
-		"host", stream.Conn().LocalPeer(),
-		"peer", stream.Conn().RemotePeer(),
-	)
-
 	// parse message and exit on error
 	msg, err := parseMessage(stream)
 	if err != nil {
-		log.Error("Failed to parse message", "err", err)
+		log.Error("Failed to parse message from peer", "err", err)
 		return // exit on error
 	}
 
-	if msg.GetType() == StatusMsgType {
-		// handle status message type
-		s.handleStreamStatus(stream, msg)
+	log.Trace(
+		"Received message from peer",
+		"host", stream.Conn().LocalPeer(),
+		"peer", stream.Conn().RemotePeer(),
+		"type", msg.GetType(),
+	)
+
+	if msg.GetType() != StatusMsgType {
+		s.handleMessage(stream, msg)
 	} else {
-		// handle other message types
-		s.handleStreamNonStatus(stream, msg)
+		s.handleStatusMessage(stream, msg)
 	}
 
-	// send message to core service
-	s.msgSend <- msg
 }
 
-// handleStreamStatus handles status messages written to the stream
-func (s *Service) handleStreamStatus(stream network.Stream, msg Message) {
+// handleMessage handles non-status messages written to the stream
+func (s *Service) handleMessage(stream network.Stream, msg Message) {
+
+	// check if status exchange is disabled or peer status is confirmed
+	if s.host.noStatus || s.host.peerStatus[stream.Conn().RemotePeer()] {
```

Comment:
I think this clause is incorrect

Reply:
how so?

Reply:
If status exchange is disabled (we do not care if we have confirmed the peer status) or the peer status has been confirmed (we care if we have confirmed the peer status and we are checking to make sure it has been confirmed) then we will send the message to the core service.